### DISCUSSION TITLE
Dynamic thresholding percentile

### DISCUSSION THREAD

**Discussion by oconnoob** [lucidrains/imagen-pytorch#122] [Ideas]
https://github.com/lucidrains/imagen-pytorch/blob/7d31add682a289278e2fcda098049e8281e1eca2/imagen_pytorch/imagen_pytorch.py#L1696

[Explanation here](https://www.assemblyai.com/blog/how-imagen-actually-works/#large-guidance-weight-samplers)

**Comment by jacobwjs** [comment]
It seems to have quite an impact on my results btw. Something to take into consideration before any serious training.

**Reply by oconnoob** [reply]
@jacobwjs thanks for that info - I can definitely see how it would have a big effect on training. Have you found the value of `0.9` to be optimal? How far away from this value did you have to go before results got poor?

Or do you mean dynamic thresholding in general is important for good results

**Reply by jacobwjs** [reply]
@oconnoob That's something I've spent little time rigorously flushing out to be honest. Best would be to do a sweep over various values. The few that I've tested align with the paper (of course).
- static thresholding definitely leads to higher contrast
- dynamic thresholding with Imagen seems (visually) to perform better with higher values (e.g. 0.95 or higher)

I recommend dynamic thresholding, but choosing values on the higher end.

I've only done a few tests, but would love to hear your results if you do as well.
